text/1180971/async_afek_gafni.tex

```diff
@@ -0,0 +1,181 @@
+    \item LEADER - the leader was found! The node send that information to every node with his id. Every other node will now send END message to the leader and prepare to end the program.
+    \item END - if the node is not the leader then he sends this messages to the leader informing leader that he is ready to end the program (he will receive messages but it will discard them). If leader gets n - 1 END messages (from every neighbour) then he sends END message to all nodes and then the program ends. It is needed to have two kinds of messages because if there was only END message (if node receive it - ends) then there would be a possibility that some node would send on closed channel.
+\end{itemize}
+
+\section*{Implementation details}
+First of all algorithm above assume that channels are not blocking. However in our implementation the node can only send one message without blocking the channel. If it sends two to the same channel the node waits until the first message is read.
+
+So the first change we implement is that every ordinary process wait for response from candidate father. We need to change also candidate process, Now candidate on response to ASK message from ordinary process will return one of three values:
+\begin{enumerate}
+    \item DEAD 
+    \item ANSWER\_ACCEPT\_CND 
+    \item ANSWER\_DENY\_CND
+\end{enumerate}
+
+\section*{Proof of correctness}
+The pair (level, id) is always increasing so there will always be the maximum candidate.
+
+\section*{Proof of complexity}
+\begin{lemma}
+For any given \(k\), the number of candidates that own \(\frac{n}{k}\) or more nodes is at most \(k\). 
+\end{lemma}
+\begin{proof}
+Let \(C_1\) and \(C_2\) be any two candidates which owned \(\frac{n}{k}\) nodes at some point in time. \(C1\) and \(C2\) must have owned disjoint sets of at least \(\frac{n}{k}\) nodes each. If they never tried to claim a node from each other, we are
+done. The first time that \(C_1\) (without loss of generality) tries to claim a node, say \(v\),
+from \(C_2\), either one of them dies, or \(C_2\) has already been killed. If \(C_1\), without loss
+of generality, caused the death of \(C_2\), then clearly it must have owned at least \(\frac{n}{k}\)
+nodes disjoint from \(C_2\), at the time of killing. If \(C_2\) is already dead, \(C_1\) must still own
+at least \(\frac{n}{k}\) nodes in order to claim \(v\) to itself. Remember that level is the number of nodes that candidate captures.
+\end{proof}
+\begin{corollary}
+The largest candidate to be killed by another candidate owns at most \(\frac{n}{2}\) nodes; the next largest owns at most \(\frac{n}{3}\) nodes.
+\end{corollary}
+\begin{lemma}
+The message complexity of Algorithm B is \(4n\) \( \ln{n} \) \(=\) \(2,773n\) \( \log_2{n}\) messages plus \(\mathcal{O}(n)\) ending sequence messages.
```

Comment:
Please put $4n \ln{n} = 2.773 n \log_2{n}$ in one bracket.

Reply:
Done